File: src/utils/llm.js
const config = require('../lib/config');
const Cerebras = require('@cerebras/cerebras_cloud_sdk'); // Import Cerebras SDK

// Initialize Cerebras Client using API key from config/env
const cerebras = new Cerebras({
  apiKey: config.cerebras.apiKey // Use the key from environment via config
});

/**
 * Function to handle chat-like conversations with Cerebras using the SDK.
 * Note: The SDK might primarily support streaming. This function will collect the stream.
 * @param {Array<object>} messages - Array of message objects ({ role: 'user'/'assistant'/'system', content: '...' })
 * @param {object} options - Optional parameters (model, max_tokens, temperature, top_p)
 * @returns {Promise<string>} - The complete assistant response string.
 */
async function cerebrasChat(messages, options = {}) {
  if (!config.cerebras.apiKey) {
    throw new Error('Cerebras API key not configured.');
  }

  try {
    const stream = await cerebras.chat.completions.create({
      messages: messages,
      model: options.model || 'llama-3.3-70b', // Default model from your example
      stream: true, // SDK seems oriented towards streaming
      max_completion_tokens: options.max_tokens || options.maxTokens || 2048, // Allow different naming conventions
      temperature: options.temperature || 0.2,
      top_p: options.top_p || 1
    });

    let fullResponse = '';
    for await (const chunk of stream) {
      fullResponse += chunk.choices[0]?.delta?.content || '';
    }
    
    console.log('Cerebras chat completion successful.');
    return fullResponse.trim();

  } catch (error) {
    console.error('Error calling Cerebras SDK:', error);
    throw new Error(`Failed to get chat completion from Cerebras: ${error.message}`);
  }
}

// REMOVE old axios-based functions if they exist
// async function cerebrasCompletion(...) { ... }

module.exports = {
  // Only export the new chat function
  cerebrasChat
}; 

File: backend/app/utils/llm.py
"""
LLM Utility - Cerebras Chat Client with consistent response handling.
"""

from cerebras.cloud.sdk import Cerebras
from app.config.settings import get_settings
from dataclasses import dataclass


settings = get_settings()

# Initialize Cerebras client
_client = None


def _get_client() -> Cerebras:
    global _client
    if _client is None:
        if not settings.cerebras_api_key:
            raise RuntimeError("CEREBRAS_API_KEY not configured")
        _client = Cerebras(api_key=settings.cerebras_api_key)
    return _client


@dataclass
class ChatResponse:
    """Response from the LLM."""
    text: str
    model: str = ""
    usage: dict | None = None


async def cerebras_chat(
    messages: list[dict],
    model: str = "llama-3.3-70b",
    temperature: float = 0.2,
    max_tokens: int = 2048,
    top_p: float = 1.0,
) -> ChatResponse:
    """
    Send messages to Cerebras and collect the streamed response.

    Args:
        messages: List of dicts with 'role' and 'content'
        model: Model name
        temperature: Sampling temperature
        max_tokens: Max completion tokens
        top_p: Top-p sampling

    Returns:
        ChatResponse with full text
    """
    client = _get_client()

    stream = client.chat.completions.create(
        messages=messages,
        model=model,
        stream=True,
        max_completion_tokens=max_tokens,
        temperature=temperature,
        top_p=top_p,
    )

    full_response = ""
    for chunk in stream:
        delta = chunk.choices[0].delta
        if delta and delta.content:
            full_response += delta.content

    return ChatResponse(text=full_response.strip(), model=model)
